src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs

```diff
@@ -312,6 +261,57 @@ private bool IsTypeExcludedFromValidation(IList<IExcludeTypeValidationFilter> fi
             return filters.Any(filter => filter.IsTypeExcluded(type));
         }
 
+        private IList<IModelValidator> ExpandValidationNodeAndGetValidators(
+            ValidationContext context, ModelExplorer modelExplorer)
+        {
+            var validationNode = context.ValidationNode;
+            if (validationNode.ChildNodes.Count != 0 ||
+                !validationNode.ValidateAllProperties ||
+                validationNode.Model == null)
+            {
+                return null;
+            }
+
+            var enumerableModel = modelExplorer.Model as IEnumerable;
+            if (enumerableModel == null)
+            {
+                foreach (var property in validationNode.ModelMetadata.Properties)
+                {
+                    var propertyExplorer = modelExplorer.GetExplorerForProperty(property.PropertyName);
+                    var propertyBindingName = property.BinderModelName ?? property.PropertyName;
+                    var childKey = ModelNames.CreatePropertyModelName(validationNode.Key, propertyBindingName);
+                    var childNode = new ModelValidationNode(childKey, property, propertyExplorer.Model)
+                    {
+                        ValidateAllProperties = true
+                    };
+                    validationNode.ChildNodes.Add(childNode);
+                }
+
+                // returning null here will ensure a new set of model validators are fetched for each child node.
+                return null;
+            }
+            else
+            {
+                var elementType = GetElementType(enumerableModel.GetType());
+                var elementMetadata = _modelMetadataProvider.GetMetadataForType(elementType);
+                var index = 0;
+                foreach (var element in enumerableModel)
+                {
+                    var elementExplorer = new ModelExplorer(_modelMetadataProvider, elementMetadata, element);
+                    var elementKey = ModelNames.CreateIndexModelName(validationNode.Key, index);
+                    var childNode = new ModelValidationNode(elementKey, elementMetadata, elementExplorer.Model)
+                    {
+                        ValidateAllProperties = true
+                    };
+
+                    validationNode.ChildNodes.Add(childNode);
+                    index++;
+                }
+
+                return GetValidators(context.ModelValidationContext.ValidatorProvider, elementMetadata);
+            }
+        }
```

Comment:
The contract of this method is really strange, and only exists because of an optimization in WebAPI2 that was done to avoid re-creating the validators for elements of a collection.

My suggestion is to just fork the code path earlier - if a few lines of boilerplate code get duped, that's not the worst thing in the world.